Fix cat forward-AD tests

qqaatw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99596

Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043

Edited: Found the skipper and xfailed some failures, which are pre-existing and unrelated to the fix in question. (Those failures are of gradgrad check, while the fix is of forward-AD).

qqaatw committed:
Fix cat forward-AD with empty tensor concatenated

[ghstack-poisoned]

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043



[ghstack-poisoned]

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043



[ghstack-poisoned]

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043



[ghstack-poisoned]

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043



[ghstack-poisoned]

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043



[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/common_utils.py:
```diff
@@ -4368,8 +4368,6 @@ def _skip_helper(self, op, device, dtype):
             self.skipTest("Skipped! Op doesn't support autograd for this dtype.")
         if not op.supports_autograd and not op.supports_forward_ad:
             self.skipTest("Skipped! autograd not supported.")
-        if op.name == "cat":
```
Looks like this was skipping all gradient tests for cat, good catch lol

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16126,7 +16126,12 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
-               DecorateInfo(unittest.expectedFailure, 'TestNNCOpInfo', 'test_nnc_correctness'),)),
+               DecorateInfo(unittest.expectedFailure, 'TestNNCOpInfo', 'test_nnc_correctness'),
+               # Previously skipped here, changed to xfail:
+               # https://github.com/pytorch/pytorch/blob/e63c502baa4a6f2109749984be701e722b3b7232/torch/testing/_internal/common_utils.py#L4371-L4372
```
nit: We probably don't need a comment about what changed (previously skipped where), could you file an issue about what sample input this is failing for? (we could put the comment about all the tests beign skipped previously there instead)

qqaatw replied:
Track here: #99806

qqaatw replied:
Replaced the comment with the issue link

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -5737,24 +5737,33 @@ Tensor lu_unpack_backward(
 
-    out_fw_grad = at::cat(fw_grads, dim);
+    out_fw_grad = fw_grads.size() > 0 ? at::cat(fw_grads, dim)
+                                      : at::_efficientzerotensor({0}, options);
```
What is the behavior here?
- we always take the tensor options of the last tensor?
- what happens for forward normally, we should probably mirror that behavior there

qqaatw replied:
1. The first forward grad defined tensor (please see above).
2. The logic in forward is a bit more complex but basically similar:

https://github.com/pytorch/pytorch/blob/7a8d0ccddff9480b3ca86bfed2f77c5ad43f35b2/aten/src/ATen/native/TensorShape.cpp#L258-L265

where `cat_should_skip_tensor` checks whether the tensor is empty.

qqaatw replied:
The difference is that the forward uses the first non-empty tensor's option, while the jvp uses the first tensor with forward grad defined (certainly empty in this very situation).

soulitzer reviewed (commented):
Thanks! Just had some minor comments.

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043

Edited: Found the skipper and xfailed some failures, which are pre-existing and unrelated to the fix in question. (Those failures are of gradgrad check, while the fix is of forward-AD).


[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -5737,24 +5737,33 @@ Tensor lu_unpack_backward(
     }
 
-    out_fw_grad = at::cat(fw_grads, dim);
+    out_fw_grad = fw_grads.size() > 0 ? at::cat(fw_grads, dim)
```
I realized I didn't really understand the issue:
- Why can't we rely on cat forward to just to handle this again? 
- Does this happen only for the torch.tensor([]) case or does it also happen for cases where numel is zero.

I could repro with Jacobian, but not actually repro with the following:
```
import torch
from torch.autograd.functional import jacobian
import torch.autograd.forward_ad as fwAD

torch.manual_seed(420)

x = torch.randn(4, 3, 32, 32)
empty = torch.Tensor([])
out = torch.cat([x, empty])

print("OK")

with fwAD.dual_level():
  x_dual = fwAD.make_dual(x, torch.ones_like(x))
  empty_dual = fwAD.make_dual(empty, torch.ones_like(empty))
  out_dual = torch.cat([x_dual, empty_dual], dim=1)
  out_p, out_t = fwAD.unpack_dual(out_dual)
  print(out_p.shape, out_t.shape)
```

qqaatw replied:
The reason that you can repro with Jacobian but not with regular forward-AD is that the Jacobian is vectorized.

As a result, the `x`'s size will actually become `(chunk_size, 4, 3, 32, 32)` and `empty`'s size will become `(chunk_size, 0)`, hence the runtime error (not happening in `cat_jvp` but the v-mapped function).

So cat forward can't handle it. I think it should only happen to `torch.tensor([])`; otherwise the code will complain when doing forward pass. The point is, normally cat only allows tensors with the same number of dimensions to be concatenated, but empty tensor is an exception, which can be arbitrarily concatenated.

So the fix this PR proposes is that we don't concatenate empty tensors' forward grads, so they won't cause the error when empty tensors' grads and non-empty tensors' grads are v-mapped together.

Correct me as I could be wrong.

soulitzer replied:
I don't think we should modify the forward AD implementation to appease an issue with the vmap formula (if that is the case). See my comment here - https://github.com/pytorch/pytorch/issues/94115#issuecomment-1520656819. It seems that it no longer fails with `torch.func.jacfwd` which uses the newer vmap.

qqaatw replied:
Except for the issue with vectorized Jacobian, this PR actually fixes the forward-AD tests, which failed before the PR:

```
Traceback (most recent call last):
  File "/Users/qqaatw/Forks/pytorch/test/test_ops_fwd_gradients.py", line 61, in test_forward_mode_AD
    self._forward_grad_helper(device, dtype, op, op.get_op(), is_inplace=False)
  File "/Users/qqaatw/Forks/pytorch/test/test_ops_fwd_gradients.py", line 49, in _forward_grad_helper
    call_grad_test_helper()
  File "/Users/qqaatw/Forks/pytorch/test/test_ops_fwd_gradients.py", line 46, in call_grad_test_helper
    self._grad_test_helper(device, dtype, op, variant, check_forward_ad=True, check_backward_ad=False,
  File "/Users/qqaatw/Forks/pytorch/torch/testing/_internal/common_utils.py", line 4362, in _grad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradcheck', check_forward_ad=check_forward_ad,
  File "/Users/qqaatw/Forks/pytorch/torch/testing/_internal/common_utils.py", line 4331, in _check_helper
    self.assertTrue(gradcheck(fn, gradcheck_args,
  File "/Users/qqaatw/Forks/pytorch/torch/testing/_internal/common_utils.py", line 3886, in gradcheck
    return torch.autograd.gradcheck(fn, inputs, **kwargs)
  File "/Users/qqaatw/Forks/pytorch/torch/autograd/gradcheck.py", line 1552, in gradcheck
    return _gradcheck_helper(**args)
  File "/Users/qqaatw/Forks/pytorch/torch/autograd/gradcheck.py", line 1572, in _gradcheck_helper
    _test_batched_grad_forward_ad(func, tupled_inputs)
  File "/Users/qqaatw/Forks/pytorch/torch/autograd/gradcheck.py", line 884, in _test_batched_grad_forward_ad
    result = _vmap(jvp)(torch.stack(tangents))
  File "/Users/qqaatw/Forks/pytorch/torch/_vmap_internals.py", line 223, in wrapped
    batched_outputs = func(*batched_inputs)
  File "/Users/qqaatw/Forks/pytorch/torch/autograd/gradcheck.py", line 865, in jvp
    dual_outputs = _as_tuple(func(*inputs_with_dual))
  File "/Users/qqaatw/Forks/pytorch/torch/testing/_internal/common_utils.py", line 4323, in fn
    output = op.gradcheck_wrapper(variant, *positional_args, **kwargs)
  File "/Users/qqaatw/Forks/pytorch/torch/testing/_internal/opinfo/core.py", line 766, in <lambda>
    gradcheck_wrapper: Callable = lambda op, *args, **kwargs: op(*args, **kwargs)
IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)
========================= short test summary info ==========================
FAILED [0.3031s] test/test_ops_fwd_gradients.py::TestFwdGradientsCPU::test_forward_mode_AD_cat_cpu_complex128
FAILED [0.1073s] test/test_ops_fwd_gradients.py::TestFwdGradientsCPU::test_forward_mode_AD_cat_cpu_float64
```

soulitzer replied:
I see, do you know which sample is failing. Do we have a repro for that?

qqaatw replied:
The same one: https://github.com/pytorch/pytorch/issues/99806

Yeah, so vamp again...
If you believe that the vamp in gradcheck is also problematic, then we should probably do a comprehensive review of our vmap-ad functionality for empty tensor handling...

Nevertheless, I feel the fix is benign.

soulitzer replied:
If we really want to fix old vmap + cat jvp (even though we don't maintain old vmap anymore) isn't the right fix to modify the vmap formula?

That said I'd recommend against fixing old vmap.

qqaatw replied:
My question is: I don't know where the problem with this fix is, e.g. the formula this fix brings for cat_jvp is wrong. If any, can you please point out.

Is it only because the old vmap is problematic and I'm not creating a fix that fixes both old vmap and cat_vjp simultaneously but a fix that makes cat_vjp compatible with the old vmap?

Unless we start switching Jacobian and forward-ad tests to the new vmap from now and drop the old one, the problem is still there, isn't it?

I would feel strange if we just suggest people not use Jacobian and skip the batch tests. (Unless the old vmap usage is fully deprecated)

I do agree with you against fixing the old vmap in favor of the new one.

soulitzer replied:
For linear functions like cat, the jvp should behave identically to the forward, this change makes cat an exception. So this adds complexity to fix an edge case for old vmap something which we don't really care about supporting.

> I would feel strange if we just suggest people not use Jacobian and skip the batch tests. (Unless the old vmap usage is fully deprecated)

Both old vmap and Jacobian (when vectorize=True) are documented as experimental features, so that should be the expected level of support. For Jacobian users, there is a clear workaround as well. 

Users won't run into the batch test issue since user-facing gradcheck has that flag disabled by default.

qqaatw replied:
I see, that makes sense. I've updated this PR and edited the skipped batch test in #66357.

albanD reviewed (commented):
I would agree with @soulitzer here that we don't want to fix the old vmap impl.
My understanding is that we only want to set `check_batched_forward_grad=False` on the cat OpInfo (on top of removing the blanket skip!). That should make sure that we test everything but the old vmap impl.

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043

Edited: Found the skipper and xfailed some failures, which are pre-existing and unrelated to the fix in question. (Those failures are of gradgrad check, while the fix is of forward-AD).


[ghstack-poisoned]

qqaatw committed:
Update on "Fix cat forward-AD with empty tensor concatenated"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043

Edited: Found the skipper and xfailed some failures, which are pre-existing and unrelated to the fix in question. (Those failures are of gradgrad check, while the fix is of forward-AD).


[ghstack-poisoned]

qqaatw commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -5787,7 +5787,7 @@ Tensor block_diag_jvp(at::TensorList tensors) {
 Tensor stack_jvp(at::TensorList tensors, int64_t dim) {
   // Basically copy of cat_jvp above
-  // TOD0: consolidate with the logic of cat_jvp
+  // TODO: consolidate with the logic of cat_jvp
```
I guess this should be `TODO` not `TOD0` :P

qqaatw committed:
Update on "Fix cat forward-AD tests"


Fixes #94115

Not sure where to add the test. There is an existing sample input but apparently doesn't fail any test.

https://github.com/pytorch/pytorch/blob/6580b160d35a75d5ceebf376d55422376d0c0d2c/torch/testing/_internal/common_methods_invocations.py#L2043

Edited: Found the skipper and xfailed some failures, which are pre-existing and unrelated to the fix in question. (Those failures are of gradgrad check, while the fix is of forward-AD).


[ghstack-poisoned]

soulitzer reviewed (approved):
Thanks!

soulitzer commented:
@pytorchbot merge
